### PR TITLE
feat(utils): added lru cache [SIG-365]

### PR DESCRIPTION
**Pull request by kedrzu**
# [SIG-365] Submodule changes

This PR contains changes to the nzyme submodule.

---

> [!NOTE]
> **Low Risk**
> Mostly additive utilities and test helpers; the only behavior change is consolidating test log capture/forwarding, which should be low impact outside tests.
> 
> **Overview**
> Adds a new `createLruCache` utility (Map-backed) to `@nzyme/utils`, exports it from the package, and includes a comprehensive Bun test suite covering eviction and recency edge cases.
> 
> Refactors test logging helpers by introducing `createTestLoggerTransport` (and `createTestLogger`) in `@nzyme/logging`, exporting them publicly, and updating `createTestContainer` in `ioc-testing` to use the shared transport instead of duplicating capture/console-forward logic.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 78638c7c5819e7ccf4ce69e8cea13dd3338aa162. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/utils/src/createLruCache.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        values: () => cache.values(),
+        keys: () => cache.keys(),
+        get size() {
+            return cache.size;
+        },
+        cache,
+    };
+
+    function get(key: K): V | undefined {
+        if (!cache.has(key)) {
```
Comment (kedrzu) [inline packages/utils/src/createLruCache.ts:67]: ## 🤖 Fixed

[**@chatgpt-codex-connector**](https://github.com/kedrzu/nzyme/pull/120#discussion_r3110263596):
> **<sub><sub>P2 Badge</sub></sub>  Preserve recency for cached undefined values**
> 
> Treating `cache.get(key) !== undefined` as the existence check breaks LRU semantics when a present entry stores `undefined`. In that case `get` returns the value but does not execute the delete+set recency update, so a recently accessed key can still be considered oldest and be evicted on the next insertion.
> 
> Useful? React with 👍 / 👎.

Switched `get()` to use `cache.has(key)` as the existence check before promoting the entry. Previously `value !== undefined` skipped promotion for cached `undefined` values, breaking LRU recency when `V` includes `undefined`. Added a regression test (`get promotes entries with undefined values`) that fails on the old code and passes on the new.

File: packages/utils/src/createLruCache.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        keys: () => cache.keys(),
+        get size() {
+            return cache.size;
+        },
+        cache,
+    };
+
+    function get(key: K): V | undefined {
+        const value = cache.get(key);
+        if (value !== undefined) {
```
[the remark this answers is not in the document]
Comment (kedrzu) [reply packages/utils/src/createLruCache.ts]: ## 🤖 Fixed

Switched `get()` to use `cache.has(key)` as the existence check before promoting the entry. Previously `value !== undefined` skipped promotion for cached `undefined` values, breaking LRU recency when `V` includes `undefined`. Added a regression test (`get promotes entries with undefined values`) that fails on the old code and passes on the new.

File: packages/utils/src/createLruCache.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    function set(key: K, value: V): void {
+        // If key already exists, delete first to update insertion order
+        cache.delete(key);
+
+        if (cache.size >= maxSize) {
+            // Evict the oldest (first) entry — iterate keys() to handle
+            // cases where `undefined` itself is a valid key.
+            for (const oldest of cache.keys()) {
```
Comment (kedrzu) [inline packages/utils/src/createLruCache.ts:84]: ## 🤖 Fixed

[**@chatgpt-codex-connector**](https://github.com/kedrzu/nzyme/pull/120#discussion_r3110263598):
> **<sub><sub>P2 Badge</sub></sub>  Evict oldest entry when oldest key is undefined**
> 
> The eviction guard checks `oldest !== undefined`, but `undefined` is a valid `Map` key. If the least-recently-used key is actually `undefined`, eviction is skipped and the subsequent `set` grows the cache past `maxSize`, violating the advertised capacity limit.
> 
> Useful? React with 👍 / 👎.

Replaced `cache.keys().next().value !== undefined` with a `for (const oldest of cache.keys()) { cache.delete(oldest); break; }` loop so that a legitimately-`undefined` oldest key is still evicted. Added a regression test (`evicts oldest entry when oldest key is undefined`) covering `LruCache<undefined | string, number>`.

File: packages/utils/src/createLruCache.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+    function set(key: K, value: V): void {
+        // If key already exists, delete first to update insertion order
+        cache.delete(key);
+
+        if (cache.size >= maxSize) {
+            // Evict the oldest (first) entry
+            const oldest = cache.keys().next().value;
+            if (oldest !== undefined) {
+                cache.delete(oldest);
```
[the remark this answers is not in the document]
Comment (kedrzu) [reply packages/utils/src/createLruCache.ts]: ## 🤖 Fixed

Replaced `cache.keys().next().value !== undefined` with a `for (const oldest of cache.keys()) { cache.delete(oldest); break; }` loop so that a legitimately-`undefined` oldest key is still evicted. Added a regression test (`evicts oldest entry when oldest key is undefined`) covering `LruCache<undefined | string, number>`.

File: packages/utils/src/createLruCache.ts (hunk 1 above)
Comment (kedrzu) [inline packages/utils/src/createLruCache.ts:67]: ## 🤖 Fixed

[**@cursor**](https://github.com/kedrzu/nzyme/pull/120#discussion_r3110272295):
> ### LRU `get` fails to promote entries with `undefined` values
> 
> **Low Severity**
> 
> 
> The `get` function checks `value !== undefined` to decide whether to promote an entry. When `V` includes `undefined` (e.g., `LruCache<string, number | undefined>`), a stored `undefined` value won't be promoted to the most recent position, silently breaking LRU ordering. Using `cache.has(key)` would be correct. The same `!== undefined` pattern in `set`'s eviction check on `oldest` would also skip eviction if `K` ever includes `undefined`.
> 
> 
> 
> 
> 
> <details>
> <summary>Additional Locations (1)</summary>
> 
> - [`packages/utils/src/createLruCache.ts#L81-L85`](https://github.com/kedrzu/nzyme/blob/a4caf7970b30bc4a2cb571ee5b2a56599ed87d95/packages/utils/src/createLruCache.ts#L81-L85)
> 
> </details>
> 
> <div><a href="https://cursor.com/open?data=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImJ1Z2JvdC12MiJ9.eyJ2ZXJzaW9uIjoxLCJ0eXBlIjoiQlVHQk9UX0ZJWF9JTl9DVVJTT1IiLCJkYXRhIjp7InJlZGlzS2V5IjoiYnVnYm90OjBlODU3NDQ2LWNlMGMtNGI4MC04NTM1LWM2MDU2ZjY0YjcxNSIsImVuY3J5cHRpb25LZXkiOiJuS2s2NVR2RXMtYXk2WTY3em85clVrUkdBc01ZTHdPNzVoU2VxRmJnTk9rIiwiYnJhbmNoIjoiZmVhdHVyZS9zaWctMzY1LXJlZmFjdG9yLXVzZWNoYXQtdG8tYS1jaGF0c3RvcmUtc2VydmljZSIsInJlcG9Pd25lciI6ImtlZHJ6dSIsInJlcG9OYW1lIjoibnp5bWUifSwiaWF0IjoxNzc2Njg0NTAwLCJleHAiOjE3NzkyNzY1MDB9.U9bPJNpuaXBFnwGE1E056LqEl2S195RvGEH4G1e27lj5PLgq1USiZP50WnUS_CMEGVXDw9oCYdmCdh2sT0JI0KbFC-DRLjiS6Pm8hvyuVKpH_wrlTw9FtgkOY1kidnOcu01dtMBICMIAgXzkj171murYoNdWDUNGKjZHUYl3RDU8pMq5gPoN1ks7QSpqlZ-6KjNmwQSwUhVo8Y0I3hh-d59ypAeyNnlg0dtuUnlTkpy_SrLaNpdih7hEXiGLOmKX3jrbn9bJDaiHD94MfU76otVz4GfzNcjV7UqBjrEwKWeibdxWuTM9FKBjhplT7gtX-DBZraFjBXH-bKz-ZHv9TA" target="_blank" rel="noopener noreferrer"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/fix-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/fix-in-cursor-light.png">Fix in Cursor</picture></a>&nbsp;<a href="https://cursor.com/agents?data=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImJ1Z2JvdC12MiJ9.eyJ2ZXJzaW9uIjoxLCJ0eXBlIjoiQlVHQk9UX0ZJWF9JTl9XRUIiLCJkYXRhIjp7InJlZGlzS2V5IjoiYnVnYm90OjBlODU3NDQ2LWNlMGMtNGI4MC04NTM1LWM2MDU2ZjY0YjcxNSIsImVuY3J5cHRpb25LZXkiOiJuS2s2NVR2RXMtYXk2WTY3em85clVrUkdBc01ZTHdPNzVoU2VxRmJnTk9rIiwiYnJhbmNoIjoiZmVhdHVyZS9zaWctMzY1LXJlZmFjdG9yLXVzZWNoYXQtdG8tYS1jaGF0c3RvcmUtc2VydmljZSIsInJlcG9Pd25lciI6ImtlZHJ6dSIsInJlcG9OYW1lIjoibnp5bWUiLCJwck51bWJlciI6MTIwLCJjb21taXRTaGEiOiJhNGNhZjc5NzBiMzBiYzRhMmNiNTcxZWU1YjJhNTY1OTllZDg3ZDk1IiwicHJvdmlkZXIiOiJnaXRodWIifSwiaWF0IjoxNzc2Njg0NTAwLCJleHAiOjE3NzkyNzY1MDB9.BeY5kmSWhc67EU8bHduok5zlwkDrTwXzcc8gzoLlibHhBnGnn7nxZdldGTAGkWRXThMRY9MnL1_8G8poy2ZHuDO17WveegzUFrlJ1nbabpUKAGhm6BqdPXbp1d_nTAiuKvgct7EaDILabwrhLDto75uMFOP9RJBCY7nE-09UB1GfzYMOf_d7TGeIHJGtMN_x_d57ftF9w1E5XsN1YK7BXGGz_20uSrd-fi-HbVxrUbUxrhNgV3XlGFuD4EbgmDgzvL7nGLiBsBCnpZf6cUabKj1ASgzlDPWUB0iRbimN0Ug61ga3ArfNB5Nd775TuX8bAjmDpQFSos-08rVokxv0cQ" target="_blank" rel="noopener noreferrer"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/fix-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/fix-in-web-light.png">Fix in Web</picture></a></div>
> 
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a4caf7970b30bc4a2cb571ee5b2a56599ed87d95. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>
> 

`get()` now uses `cache.has(key)` for presence detection, so entries whose value is `undefined` are still promoted to the most-recently-used position. Covered by a new test `get promotes entries with undefined values`.

File: packages/utils/src/createLruCache.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+    function get(key: K): V | undefined {
+        const value = cache.get(key);
+        if (value !== undefined) {
+            // Move to most recent by re-inserting
+            cache.delete(key);
+            cache.set(key, value);
+        }
+        return value;
+    }
```
[the remark this answers is not in the document]
Comment (kedrzu) [reply packages/utils/src/createLruCache.ts:75]: ## 🤖 Fixed

`get()` now uses `cache.has(key)` for presence detection, so entries whose value is `undefined` are still promoted to the most-recently-used position. Covered by a new test `get promotes entries with undefined values`.

File: packages/utils/src/index.ts [hunk 6]
```diff
@@ -42,6 +42,7 @@ export * from './createCache.js';
 export * from './createEventEmitter.js';
 export * from './createExponentialBackoff.js';
 export * from './createGetter.js';
+export * from './createLruCache.js';
```
Comment (kedrzu) [inline packages/utils/src/index.ts:45]: ## 🤖 Fixed

[**@cursor**](https://github.com/kedrzu/nzyme/pull/120#discussion_r3110272288):
> ### New `createLruCache` not exported from package barrel file
> 
> **Medium Severity**
> 
> 
> `createLruCache` is defined and exported from its own module but is not re-exported from `packages/utils/src/index.ts`. Every other utility in the package is included in the barrel file. Consumers importing from the package entry point (e.g., `@nzyme/utils`) won't have access to this new function.
> 
> 
> 
> 
> 
> <div><a href="https://cursor.com/open?data=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImJ1Z2JvdC12MiJ9.eyJ2ZXJzaW9uIjoxLCJ0eXBlIjoiQlVHQk9UX0ZJWF9JTl9DVVJTT1IiLCJkYXRhIjp7InJlZGlzS2V5IjoiYnVnYm90OjA3NWI0NzNhLTFlMDgtNGQxYS04MTExLTA1Y2FhNGRlMzliOCIsImVuY3J5cHRpb25LZXkiOiJPZEk5SkhlbjZIcjVwOS1ycFNEdGxuMGNaMmRvaVRpXy1wd3BjQy12LW1RIiwiYnJhbmNoIjoiZmVhdHVyZS9zaWctMzY1LXJlZmFjdG9yLXVzZWNoYXQtdG8tYS1jaGF0c3RvcmUtc2VydmljZSIsInJlcG9Pd25lciI6ImtlZHJ6dSIsInJlcG9OYW1lIjoibnp5bWUifSwiaWF0IjoxNzc2Njg0NTAwLCJleHAiOjE3NzkyNzY1MDB9.dz0XdNxghJSSqbSYAtyAX4llNqwNf8KpPsSIR8BESa8-O77ouTpYg96vnn0cfXubo18TkeiPI-SmmLUV19uVMQHE8lb3pKGsro_vpsL-bwFyLWOlx8WcE0F2lJeRITIbex_kyRowBzrvYh1hQ6zWumYxODtyR1_FJ_bV5PtLIWi3idBwvvTzoNx4GFB-0Po0Vi3gonc_2kNnQY1F1goiEwZuwWNRT-A92wuE47gRKtkrE_YcXqglGpx24Scn_hf-QsAnnk7Gq5FOhhaSuNAeZpT6T-hdUMzCBqeeTwpBe22gtfFPTq_M0HKqOcuUcYOIg4tIqC6pri8Wkyw-_4H8OQ" target="_blank" rel="noopener noreferrer"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/fix-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/fix-in-cursor-light.png">Fix in Cursor</picture></a>&nbsp;<a href="https://cursor.com/agents?data=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImJ1Z2JvdC12MiJ9.eyJ2ZXJzaW9uIjoxLCJ0eXBlIjoiQlVHQk9UX0ZJWF9JTl9XRUIiLCJkYXRhIjp7InJlZGlzS2V5IjoiYnVnYm90OjA3NWI0NzNhLTFlMDgtNGQxYS04MTExLTA1Y2FhNGRlMzliOCIsImVuY3J5cHRpb25LZXkiOiJPZEk5SkhlbjZIcjVwOS1ycFNEdGxuMGNaMmRvaVRpXy1wd3BjQy12LW1RIiwiYnJhbmNoIjoiZmVhdHVyZS9zaWctMzY1LXJlZmFjdG9yLXVzZWNoYXQtdG8tYS1jaGF0c3RvcmUtc2VydmljZSIsInJlcG9Pd25lciI6ImtlZHJ6dSIsInJlcG9OYW1lIjoibnp5bWUiLCJwck51bWJlciI6MTIwLCJjb21taXRTaGEiOiJhNGNhZjc5NzBiMzBiYzRhMmNiNTcxZWU1YjJhNTY1OTllZDg3ZDk1IiwicHJvdmlkZXIiOiJnaXRodWIifSwiaWF0IjoxNzc2Njg0NTAwLCJleHAiOjE3NzkyNzY1MDB9.mmWSfO3SqVJ6E9mh__6WCYAJKBut2aVm0sep77G4x9P1iKFoCIs5kjwRqpKNiXzF7y80rkgXta76dJ8iZiHwatmOZKt5NQ7bkcq38NXzhRPE23WU6EuY6Ezpzi2VTNzWoSMwrJ54U0Ow-O75BBZvMuTdKO4vWI4n74-CLLIwrehVvl9ofEXejdgxwYg5KAfh3L2PDMwAdMbwFlev7_M2rG25oQwIiXTlyE7UJbZMcR18qYXvxsnjBsW70_p29x-TrUJSPGBBYzfaVfcoQ8en7dFfwd0VHJwG0Fz1XO6Dtjq8SbjRwCtGqtlAXO-f-3cqqSy1tnmQY_tAtzW4E1Kl5w" target="_blank" rel="noopener noreferrer"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/fix-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/fix-in-web-light.png">Fix in Web</picture></a></div>
> 
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a4caf7970b30bc4a2cb571ee5b2a56599ed87d95. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>
> 

Added `export * from './createLruCache.js';` to the barrel at line 45, placed alphabetically between `createGetter` and `createMemo`. Verified with `bun tsc --build` (passes). The existing consumer in the sigma repo (`packages/client/src/chat/ChatStore.ts`) imports directly from `@nzyme/utils/createLruCache.js`, which is the granular-import convention used throughout that file, so no consumer changes were needed — the barrel export just makes the function available for consumers that use the barrel import.

File: packages/utils/src/createLruCache.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    readonly cache: Map<K, V>;
+}
+
+/**
+ * Creates an LRU cache backed by a Map.
+ * Map preserves insertion order — delete+set moves an entry to the end (most recent).
+ *
+ * @__NO_SIDE_EFFECTS__
+ */
+export function createLruCache<K, V>(options: CreateLruCacheOptions<K, V>): LruCache<K, V> {
```
[the remark this answers is not in the document]
Comment (kedrzu) [reply packages/utils/src/createLruCache.ts:48]: ## 🤖 Fixed

Added `export * from './createLruCache.js';` to the barrel at line 45, placed alphabetically between `createGetter` and `createMemo`. Verified with `bun tsc --build` (passes). The existing consumer in the sigma repo (`packages/client/src/chat/ChatStore.ts`) imports directly from `@nzyme/utils/createLruCache.js`, which is the granular-import convention used throughout that file, so no consumer changes were needed — the barrel export just makes the function available for consumers that use the barrel import.
